### PR TITLE
feat(di): DI Abstract Contract Layer — di/core.py ABCs and Pure Functions (#905)

### LINKED ISSUE #905: DI – Abstract Contract Layer: di/core.py ABCs and Pure Functions
# Technical Requirements Document: DI – Abstract Contract Layer: di/core.py ABCs and Pure Functions

**Project:** Tiferet Framework
**Repository:** https://github.com/greatstrength/tiferet
**Date:** July 19, 2026
**Version:** Core DDD Parity IV – DI Redesign and Events Architecture (tracking milestone)

## 1. Overview

The `tiferet/di/` package currently consolidates all dependency-injection functionality in `settings.py` — a single module that mixes abstract contracts, pure helper functions, and concrete engine-specific implementation code. This story extracts the abstract contract layer into a dedicated `di/core.py` module, keeping it **domain-only and engine-free**: it imports only stdlib and `domain.ServiceDependency`, with no reference to `dependency-injector` or assets. The two pure helpers (`injectable_parameter_names`, `normalize_flags`) move from `settings.py` to `core.py` as the canonical source of truth. Two new ABCs — `ServiceContainer` and `ServiceResolver` — define the typed interfaces that all concrete DI implementations must satisfy. `ServiceResolver` provides a template-method `get_dependency` that manages a per-flag container cache and delegates construction to the abstract `build_container`. `di/__init__.py` is updated to re-export the new symbols alongside the existing `settings.py` exports, and a `tests/di/test_core.py` test module is added.

### In Scope
- Add `tiferet/di/core.py` with `ServiceContainer` ABC, `ServiceResolver` ABC, `injectable_parameter_names`, and `normalize_flags`.
- Update `tiferet/di/__init__.py` to export `ServiceContainer`, `ServiceResolver`, `injectable_parameter_names`, and `normalize_flags` from `.core`.
- Add `tests/di/test_core.py` covering both ABCs and both pure functions.

### Out of Scope
- Concrete implementations using `dependency-injector` — Story 8.
- Updates to `di/settings.py` or retirement of `di/dynamic.py` — Story 9.
- Any consumer changes in blueprints or contexts.

## 3. Components Affected

| Component | File/Path | Changes |
|-----------|-----------|---------|
| DI core module | `tiferet/di/core.py` | Add: `injectable_parameter_names`, `normalize_flags`, `ServiceContainer` (ABC), `ServiceResolver` (ABC) |
| DI package | `tiferet/di/__init__.py` | Update: add `ServiceContainer`, `ServiceResolver`, `injectable_parameter_names`, `normalize_flags` to exports from `.core` |
| DI core tests | `tests/di/test_core.py` | Add new test module |

### 4.1 `di/core.py` — module structure

Module docstring: `"""Tiferet DI Core"""`

```python
# *** imports

# ** core
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Tuple
import inspect

# ** app
from ..domain import ServiceDependency

# *** functions

# ** function: injectable_parameter_names
# ** function: normalize_flags

# *** classes

# ** class: service_container
# ** class: service_resolver
```

This module has **no dependency on `dependency-injector`** and **no import of assets**. It is domain-only.

### 4.2 `injectable_parameter_names` — pure function

```python
# ** function: injectable_parameter_names
def injectable_parameter_names(service_type: type) -> List[str]:
    '''
    Return the injectable constructor parameter names for a service type.

    Excludes ``self`` and variadic (``*args`` / ``**kwargs``) parameters.
    Types whose constructor cannot be inspected are treated as no-arg.

    :param service_type: The service class to inspect.
    :type service_type: type
    :return: The list of injectable constructor parameter names.
    :rtype: List[str]
    '''

    # Inspect the constructor signature; treat uninspectable types as no-arg.
    try:
        signature = inspect.signature(service_type.__init__)
    except (ValueError, TypeError):
        return []

    # Collect parameter names, skipping self and variadic parameters.
    names = []
    for name, parameter in signature.parameters.items():

        # Skip the bound self parameter.
        if name == 'self':
            continue

        # Skip variadic positional and keyword parameters.
        if parameter.kind in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD):
            continue

        # Record the injectable parameter name.
        names.append(name)

    # Return the collected parameter names.
    return names
```

### 4.3 `normalize_flags` — pure function

```python
# ** function: normalize_flags
def normalize_flags(*flags) -> List[str]:
    '''
    Flatten a mixed sequence of strings, lists, and tuples into a flat list of strings.

    :param flags: The flags to normalize (strings, lists, or tuples).
    :type flags: Tuple[Any, ...]
    :return: A flat list of flag strings.
    :rtype: List[str]
    '''

    # Flatten one level of lists/tuples into a single flag list, stringifying every flag.
    normalized = []
    for flag in flags:

        # Expand nested list/tuple flag groups, coercing each member to a string.
        if isinstance(flag, (list, tuple)):
            normalized.extend(str(f) for f in flag)

        # Append scalar flags coerced to strings.
        else:
            normalized.append(str(flag))

    # Return the flattened flag list.
    return normalized
```

### 4.4 `ServiceContainer` — abstract base class

`ServiceContainer` defines the typed registration and resolution contract. All concrete DI containers must implement every method. The `add_service` method accepts a `ServiceDependency` domain object (not a raw `type`) — concrete implementations are responsible for importing the service class from the dependency's `module_path` and `class_name`.

Abstract methods:

| Method | Signature | Purpose |
|--------|-----------|---------|
| `add_service` | `(service_id: str, service: ServiceDependency)` | Register a service dependency by ID |
| `add_constant` | `(constant_id: str, value: Any)` | Register a scalar constant by ID |
| `get_dependency` | `(dependency_id: str) -> Any` | Resolve a registered dependency or constant |
| `has_dependency` | `(dependency_id: str) -> bool` | Return True when a dependency is registered |
| `remove_dependency` | `(dependency_id: str)` | Deregister a dependency; idempotent |
| `load_container` | `(services: Dict[str, ServiceDependency] = None, constants: Dict[str, Any] = None)` | Bulk-load; must register constants before services |

Each method raises `NotImplementedError` in the base.

### 4.5 `ServiceResolver` — abstract base class

`ServiceResolver` manages a per-flag-tuple cache of `ServiceContainer` instances and provides a template-method `get_dependency` that builds and caches on a miss.

| Artifact | Kind | Description |
|----------|------|-------------|
| `_containers` | attribute | `Dict[Tuple[str, ...], ServiceContainer]` — per-flag container cache |
| `__init__` | init | Initialize `_containers = {}` |
| `add_container` | method | Cache a container under `tuple(normalize_flags(*flags))`; return the container |
| `get_container` | method | Return `self._containers.get(tuple(normalize_flags(*flags)))` or `None` |
| `build_container` | abstract method | Build a new `ServiceContainer` for the given `flags: List[str]` |
| `get_dependency` | method | Normalize flags; get or build-and-cache container; return `container.get_dependency(service_id)` |

`get_dependency` signature:
```python
def get_dependency(self, service_id: str, *flags) -> Any:
```

### 4.6 `di/__init__.py` — export update

Add to `__all__` and import from `.core`:
- `ServiceContainer` (new ABC — distinct from the concrete `ServiceContainer` in `.settings`)
- `ServiceResolver` (new ABC — distinct from the concrete `ServiceResolver` in `.settings`)
- `injectable_parameter_names` (moved canonical source; `.settings` version is now a duplicate)
- `normalize_flags` (same)

The existing exports from `.settings` (`ServiceProvider`, `ServiceContainer`, `ServiceResolver`, `create_cache_key`, `merge_settings`) are retained for backward compatibility.

### 4.7 Test requirements (`tests/di/test_core.py`)

Standalone test functions under `# *** tests`:

| Test name | What it verifies |
|-----------|-----------------|
| `test_injectable_parameter_names_basic` | Returns correct names for a simple `__init__` |
| `test_injectable_parameter_names_skips_self_and_variadics` | `self`, `*args`, `**kwargs` excluded |
| `test_injectable_parameter_names_uninspectable` | Returns `[]` for a C-extension type |
| `test_normalize_flags_strings` | Flat list of strings is returned unchanged |
| `test_normalize_flags_expands_list` | A list within flags is flattened |
| `test_normalize_flags_coerces_to_str` | Non-string scalars and list members become strings |
| `test_service_container_is_abstract` | Instantiating `ServiceContainer` directly raises `TypeError` |
| `test_service_resolver_is_abstract` | Instantiating `ServiceResolver` directly raises `TypeError` |
| `test_service_resolver_add_and_get_container` | `add_container` + `get_container` round-trip with normalized flags |
| `test_service_resolver_get_dependency_calls_build_on_miss` | Concrete subclass `build_container` is called when no cached container exists |

## 5. Acceptance Criteria

1. `tiferet/di/core.py` exists with a `# *** functions` section (two functions) and a `# *** classes` section (two ABCs).
2. `from tiferet.di.core import ServiceContainer, ServiceResolver, injectable_parameter_names, normalize_flags` succeeds.
3. `ServiceContainer` cannot be instantiated directly; it declares exactly the 6 abstract methods listed in §4.4.
4. `ServiceResolver` cannot be instantiated directly; `get_dependency(service_id, *flags)` builds and caches a container on the first call and reuses the cached container on subsequent calls with the same flags.
5. `injectable_parameter_names` returns parameter names excluding `self` and variadics; returns `[]` for uninspectable types.
6. `normalize_flags(['a', 'b'], 'c', 42)` returns `['a', 'b', 'c', '42']`.
7. `from tiferet.di import ServiceContainer, ServiceResolver, injectable_parameter_names, normalize_flags` succeeds (via `__init__.py`).
8. All `tests/di/test_core.py` tests pass.
9. `pytest tiferet/ tests/` passes with no regressions.

## 6. Non-Functional Requirements

- `di/core.py` has **no import from `dependency-injector`** and **no import from `..assets`** — domain-only.
- Module docstring: `"""Tiferet DI Core"""`.
- `ServiceContainer` and `ServiceResolver` use `# ** class: service_container` and `# ** class: service_resolver` artifact comments respectively.
- All abstract methods carry full RST docstrings with `:param`, `:type`, `:return`, `:rtype`.
- One empty line between each `# *` section; one empty line after docstrings.

## 7. Prerequisites

| Dependency | Status in `main` |
|------------|-----------------|
| Parity III Story 4 — `ServiceDependency` available in `domain/core.py` | Pending Parity III |

## 8. Related Code Style Documentation

- `tiferet-code-style` / [code_style.md](https://github.com/greatstrength/tiferet/blob/main/docs/core/code_style.md)
- `tiferet-code-di` / [di.md](https://github.com/greatstrength/tiferet/blob/main/docs/core/di.md)

### PR DESCRIPTION
**Pull request by Ashatz**
## Summary

Implements Parity IV Story 7 — introduces `tiferet/di/core.py` as the abstract DI contract layer.

## Changes

**`tiferet/di/core.py`** (new)
- `injectable_parameter_names` — pure function; returns injectable constructor parameter names (excludes `self` / variadics; treats uninspectable types as no-arg)
- `normalize_flags` — pure function; flattens mixed string/list/tuple flag inputs into a flat `List[str]`
- `ServiceContainer` ABC — 6 abstract methods: `add_service`, `add_constant`, `get_dependency`, `has_dependency`, `remove_dependency`, `load_container`
- `ServiceResolver` ABC — template-method `get_dependency` (builds+caches container on miss); abstract `build_container`; concrete `add_container` / `get_container` helpers with tuple-keyed per-flag cache
- Domain-only: no `dependency-injector` import, no assets import

**`tiferet/di/__init__.py`** (updated)
- Re-exports `ServiceContainer`, `ServiceResolver`, `injectable_parameter_names`, `normalize_flags` from `.core` (shadows the legacy `settings.py` names; backward compat preserved)

**`tests/di/test_core.py`** (new)
- 10 test cases covering all ACs from TRD §4.7

## Test Results

- 60 di tests pass (10 new + 50 existing)
- 46 domain/events/mappers/repos/assets/interfaces tests pass — no regressions

## Conversation

https://app.warp.dev/conversation/4f6586cd-8d6d-46eb-9e19-a253e2edb78f

Closes #905

### CODE REVIEW

**Review by Ashatz — COMMENTED** [review]
test

**Review by Ashatz — COMMENTED** [review]
## PR #924 Review — DI Abstract Contract Layer

Two deviations to address before merge. `di/core.py` itself is clean — all differences from the reference are the branch being ahead (concise docstrings, bare `raise NotImplementedError()`, required `flags` param on `build_container`). No changes needed there.

### Missing test (global finding)

**`tests/di/test_core.py` — DEVIATION: Missing `test_normalize_flags_empty`**

The reference has a dedicated test verifying `normalize_flags()` called with *no* arguments returns `[]`. The branch covers string inputs, list expansion, and coercion, but omits the zero-argument edge case. Add:

```python
# ** test: normalize_flags_empty
def test_normalize_flags_empty():
    '''
    normalize_flags returns an empty list when called with no arguments.
    '''

    # Call with no arguments.
    result = normalize_flags()

    # Result should be an empty list.
    assert result == []
```

---

Co-Authored-By: Oz <oz-agent@warp.dev>

File: tiferet/di/__init__.py [hunk 1]
```diff
@@ -23,6 +23,12 @@
     merge_settings,
 )
 from .dynamic import DynamicServiceProvider
+from .core import (
```
Comment (Ashatz) [inline tiferet/di/__init__.py:22]: **DEVIATION — Duplicate/shadowed imports**

`ServiceContainer`, `ServiceResolver`, `injectable_parameter_names`, and `normalize_flags` are now imported from both `.settings` (already present on `main`) and `.core` (lines 27–30 added here). The `.settings` imports are silently overwritten by the `.core` imports.

Fix: remove those four names from the `from .settings import (...)` block on `main`. Only keep what exclusively lives in `settings.py`:

```python
from .settings import (
    ServiceProvider,
    create_cache_key,
    merge_settings,
)
```

File: tests/di/test_core.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
+"""Tests for tiferet.di.core — ABCs and pure DI helper functions."""
```
Comment (Ashatz) [inline tests/di/test_core.py]: **DEVIATION — Module docstring format**

The framework convention is `"Tiferet <Module> Tests"`. Change to:

```python
"""Tiferet DI Core Tests"""
```

**Comment by Ashatz** [comment]
Review conversation: https://app.warp.dev/conversation/019f9174-bfff-7ce3-bc6b-dd977d6c40c5
